In [ ]:
## Forecasting Vaccine Uptake: A Data-Driven Approach

![image](H1N1.jpg)

## NoteBook Structure
1) Business Understanding<br>
2) Data Understanding<br>
3) Data Preparation<br>
4) Modeling<br>
5) Evaluation<br>
6) Conclusion<br>
7) Recommendation<br>
8) Next Steps<br>

## Step (1) Business Understanding

Vaccination is widely recognized as one of the most significant achievements in public health, having successfully eradicated or controlled numerous infectious diseases. Despite this progress, recent years have seen a troubling rise in vaccine skepticism, which has contributed to declining immunization rates and the resurgence of preventable diseases. Understanding the underlying factors driving vaccine hesitancy, as well as accurately predicting vaccination uptake, is therefore critical for public health officials and policymakers.

This project utilizes data from the National H1N1 Flu Survey (NHFS 2009) to predict individuals’ likelihood of receiving the H1N1 influenza vaccine. By analyzing historical vaccination patterns and associated demographic, behavioral, and attitudinal factors, the study seeks to provide insights into vaccination decision-making. These insights are particularly relevant in the context of emerging and ongoing global health threats, such as the COVID-19 pandemic, where vaccine acceptance plays a central role in disease control and prevention.

## Step 1.1 Problem Statement

Vaccine hesitancy represents a major challenge to public health initiatives, as it contributes to reduced immunization coverage and heightened susceptibility to infectious disease outbreaks. Gaining a deeper understanding of the factors that influence individuals’ vaccination decisions is essential for developing targeted interventions and strengthening community immunity. Within this context, this project seeks to predict individuals’ likelihood of receiving the H1N1 influenza vaccine by applying machine learning techniques to data obtained from the National H1N1 Flu Survey (NHFS 2009).

### Objectives

- Prediction: Develop and apply machine learning models to predict individuals’ likelihood of receiving the H1N1 influenza vaccine based on demographic, socio-economic, and attitudinal characteristics.

- Identification of Influential Factors: Identify and quantify the key determinants of H1N1 vaccine acceptance, including healthcare provider recommendations, health insurance status, perceptions of vaccine effectiveness, and individual risk perceptions related to H1N1 infection.

- Model Evaluation: Evaluate and compare the predictive performance of multiple machine learning algorithms—namely Decision Tree Classifier, Logistic Regression, Random Forest, K-Nearest Neighbors (KNN) Classifier, Gradient Boosting Classifier, and XGBoost Classifier—in predicting H1N1 vaccine uptake.

- Impact Analysis: Examine the practical implications of the predictive models and extract actionable insights that can inform public health strategies and policy interventions aimed at increasing vaccination coverage.

## Step 2 Data Understanding

In [1]:
# Relevant Libraries
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns
import category_encoders as ce

from matplotlib.ticker import FixedLocator

from sklearn.metrics import roc_curve, roc_auc_score
from sklearn.preprocessing import StandardScaler, MinMaxScaler, MaxAbsScaler, OneHotEncoder
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.impute import SimpleImputer

from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer



from sklearn.model_selection import train_test_split, GridSearchCV, cross_validate, cross_val_score

from sklearn.metrics import accuracy_score, recall_score, precision_score, f1_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import auc

from sklearn.dummy import DummyClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.neighbors import KNeighborsClassifier

import xgboost     # extreme gradient boosting

# To visualize the 100 columns in data
pd.options.display.max_columns=100

In [12]:
# Load the dataset and display

df= pd.read_csv('Data/H1N1_Flu_Vaccines.csv')
df.head(10)

respondent_id  h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
0              0           1.0             0.0                        0.0   
1              1           3.0             2.0                        0.0   
2              2           1.0             1.0                        0.0   
3              3           1.0             1.0                        0.0   
4              4           2.0             1.0                        0.0   
5              5           3.0             1.0                        0.0   
6              6           0.0             0.0                        0.0   
7              7           1.0             0.0                        0.0   
8              8           0.0             2.0                        0.0   
9              9           2.0             1.0                        0.0   

   behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
0                   0.0                   0.0                    0.0   
1                   1.0                   0.0                    1.0   
2                   1.0                   0.0                    0.0   
3                   1.0                   0.0                    1.0   
4                   1.0                   0.0                    1.0   
5                   1.0                   0.0                    1.0   
6                   0.0                   0.0                    0.0   
7                   1.0                   0.0                    1.0   
8                   1.0                   0.0                    1.0   
9                   1.0                   0.0                    0.0   

   behavioral_large_gatherings  behavioral_outside_home  \
0                          0.0                      1.0   
1                          0.0                      1.0   
2                          0.0                      0.0   
3                          1.0                      0.0   
4                          1.0                      0.0   
5                          0.0                      0.0   
6                          0.0                      0.0   
7                          0.0                      0.0   
8                          1.0                      1.0   
9                          1.0                      0.0   

   behavioral_touch_face  doctor_recc_h1n1  doctor_recc_seasonal  \
0                    1.0               0.0                   0.0   
1                    1.0               0.0                   0.0   
2                    0.0               NaN                   NaN   
3                    0.0               0.0                   1.0   
4                    1.0               0.0                   0.0   
5                    1.0               0.0                   1.0   
6                    0.0               0.0                   0.0   
7                    1.0               1.0                   0.0   
8                    1.0               0.0                   0.0   
9                    1.0               0.0                   0.0   

   chronic_med_condition  child_under_6_months  health_worker  \
0                    0.0                   0.0            0.0   
1                    0.0                   0.0            0.0   
2                    1.0                   0.0            0.0   
3                    1.0                   0.0            0.0   
4                    0.0                   0.0            0.0   
5                    0.0                   0.0            0.0   
6                    0.0                   0.0            0.0   
7                    1.0                   0.0            0.0   
8                    0.0                   0.0            0.0   
9                    1.0                   0.0            0.0   

   health_insurance  opinion_h1n1_vacc_effective  opinion_h1n1_risk  \
0               1.0                          3.0                1.0   
1               1.0                          5.0                4.0   
2               NaN               

In [8]:
# check the shape
df.shape

(26707, 36)

In [9]:
# check data info

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26707 entries, 0 to 26706
Data columns (total 36 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   respondent_id                26707 non-null  int64  
 1   h1n1_concern                 26615 non-null  float64
 2   h1n1_knowledge               26591 non-null  float64
 3   behavioral_antiviral_meds    26636 non-null  float64
 4   behavioral_avoidance         26499 non-null  float64
 5   behavioral_face_mask         26688 non-null  float64
 6   behavioral_wash_hands        26665 non-null  float64
 7   behavioral_large_gatherings  26620 non-null  float64
 8   behavioral_outside_home      26625 non-null  float64
 9   behavioral_touch_face        26579 non-null  float64
 10  doctor_recc_h1n1             24547 non-null  float64
 11  doctor_recc_seasonal         24547 non-null  float64
 12  chronic_med_condition        25736 non-null  float64
 13  child_under_6_mo

In [10]:
# description of data
df.describe()


respondent_id  h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
count   26707.000000  26615.000000    26591.000000               26636.000000   
mean    13353.000000      1.618486        1.262532                   0.048844   
std      7709.791156      0.910311        0.618149                   0.215545   
min         0.000000      0.000000        0.000000                   0.000000   
25%      6676.500000      1.000000        1.000000                   0.000000   
50%     13353.000000      2.000000        1.000000                   0.000000   
75%     20029.500000      2.000000        2.000000                   0.000000   
max     26706.000000      3.000000        2.000000                   1.000000   

       behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
count          26499.000000          26688.000000           26665.000000   
mean               0.725612              0.068982               0.825614   
std                0.446214              0.253429               0.379448   
min                0.000000              0.000000               0.000000   
25%                0.000000              0.000000               1.000000   
50%                1.000000              0.000000               1.000000   
75%                1.000000              0.000000               1.000000   
max                1.000000              1.000000               1.000000   

       behavioral_large_gatherings  behavioral_outside_home  \
count                  26620.00000             26625.000000   
mean                       0.35864                 0.337315   
std                        0.47961                 0.472802   
min                        0.00000                 0.000000   
25%                        0.00000                 0.000000   
50%                        0.00000                 0.000000   
75%                        1.00000                 1.000000   
max                        1.00000                 1.000000   

       behavioral_touch_face  doctor_recc_h1n1  doctor_recc_seasonal  \
count           26579.000000      24547.000000          24547.000000   
mean                0.677264          0.220312              0.329735   
std                 0.467531          0.414466              0.470126   
min                 0.000000          0.000000              0.000000   
25%                 0.000000          0.000000              0.000000   
50%                 1.000000          0.000000              0.000000   
75%                 1.000000          0.000000              1.000000   
max                 1.000000          1.000000              1.000000   

       chronic_med_condition  child_under_6_months  health_worker  \
count           25736.000000          25887.000000   25903.000000   
mean                0.283261              0.082590       0.111918   
std                 0.450591              0.275266       0.315271   
min                 0.000000              0.000000       0.000000   
25%                 0.000000              0.000000       0.000000   
50%                 0.000000              0.000000       0.000000   
75%                 1.000000              0.000000       0.000000   
max                 1.000000              1.000000       1.000000   

       health_insurance  opinion_h1n1_vacc_effective  opinion_h1n1_risk  \
count       14433.00000                 26316.000000       26319.000000   
mean            0.87972                     3.850623           2.342566   
std             0.32530                     1.007436           1.285539   
min             0.00000                     1.000000           1.000000   
25%             1.00000                     3.000000           1.000000   
50%             1.00000                     4.000000           2.000000   
75%             1.00000                     5.000000           4.000000   
max             1.00000                     5.000000           5.000000   

       opinion_h1n1_sick_from_vacc  opinion_seas_vacc_effective  \
count     

In [13]:
#checking value counts
vaccine_counts_h1n1 = df['h1n1_vaccine'].value_counts()
vaccine_counts_seasonal = df['seasonal_vaccine'].value_counts()

print("Value Counts for h1n1_vaccine:\n", vaccine_counts_h1n1)
print("\nValue Counts for seasonal_vaccine:\n", vaccine_counts_seasonal)

Value Counts for h1n1_vaccine:
 h1n1_vaccine
0    21033
1     5674
Name: count, dtype: int64

Value Counts for seasonal_vaccine:
 seasonal_vaccine
0    14272
1    12435
Name: count, dtype: int64


H1N1 Flu Vaccine:

- 21,033 respondents did not receive the H1N1 flu vaccine.
- 5,674 respondents received the H1N1 flu vaccine.
Seasonal Flu Vaccine:

- 14,272 respondents did not receive the seasonal flu vaccine.
- 12,435 respondents received the seasonal flu vaccine.

- The value counts indicate that the distribution of respondents across the "seasonal_vaccine" classes is more balanced compared to the distribution across the "h1n1_vaccine" classes.

## Step 3 EDA

In [ ]:
# We make a copy of the main dataframe so as to use the new one